File: quickSort.php
<?php
function quickSort(&$arr, $first, $last){
	if ($first < $last){
		$pivotIdx = partition($arr, $first, $last);
		quickSort($arr, $first, $pivotIdx - 1);
		quickSort($arr, $pivotIdx + 1, $last);
	}
}
function partition(&$arr, $first, $last){
	$pivot = $arr[$first];
	$left = $first + 1;
	$right = $last;
	$finished = false;
	
	while (!$finished){
		while ($left <= $right && $arr[$left] <= $pivot){
			$left++;
		}
		while ($right >= $left && $arr[$right] >= $pivot){
			$right--;
		}
		if ($right < $left){$finished = true;}
		else{
			$tmp = $arr[$left];
			$arr[$left] = $arr[$right];
			$arr[$right] = $tmp;
		}
	}
	$tmp = $arr[$first];
	$arr[$first] = $arr[$right];
	$arr[$right] = $tmp;
	return $right;
}

$arr = array(2,6,4,8,5,2,1);
quickSort($arr, 0, count($arr) - 1);
print_r($arr);
?>

File: quickSort.py
def quickSort(lst, first, last):
    if first < last:
        pivotIdx = partition(lst, first, last)
        quickSort(lst, first, pivotIdx - 1)
        quickSort(lst, pivotIdx + 1, last)

def partition(lst, first, last):
    pivot = lst[first]
    left, right = first+1, last
    finished = False

    while not finished:
        while left <= right and lst[left] <= pivot:
            left += 1
        while right >= left and lst[right] >= pivot:
            right -= 1

        if right < left:
            finished = True
        else:
            lst[left], lst[right] = lst[right], lst[left]

    lst[first], lst[right] = lst[right], lst[first]
    return right

lst = [6,3,2,5,7,7,2,3]
quickSort(lst, 0, len(lst) - 1)
print(lst)
